keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -69,40 +70,42 @@ class EnvironmentPromotionChecker(
                   constraintRunner.checkEnvironment(envContext)
 
                   // everything the constraint runner has already approved
-                  val queuedForApproval: MutableSet<String> = repository
-                    .getQueuedConstraintApprovals(deliveryConfig.name, environment.name, artifact.reference)
-                    .toMutableSet()
+                  val queuedForApproval = repository
+                    .getArtifactVersionsQueuedForApproval(deliveryConfig.name, environment.name, artifact.reference)
+                    .toMutableList()
 
                   /**
                    * Approve all constraints starting with oldest first so that the ordering is
                    * maintained.
                    */
                   queuedForApproval
-                    .sortedWith(artifact.versioningStrategy.comparator.reversed())
-                    .forEach { v ->
+                    .reversed()
+                    .forEach { artifactVersion ->
                       /**
                        * We don't need to re-invoke stateful constraint evaluators for these, but we still
                        * check stateless constraints to avoid approval outside of allowed-times.
                        */
                       log.debug(
-                        "Version $v of artifact ${artifact.name} is queued for approval, " +
+                        "Version ${artifactVersion.version} of artifact ${artifact.name} is queued for approval, " +
                           "and being evaluated for stateless constraints in environment ${environment.name}"
                       )
-                      if (constraintRunner.checkStatelessConstraints(artifact, deliveryConfig, v, environment)) {
-                        approveVersion(deliveryConfig, artifact, v, environment.name)
-                        repository.deleteQueuedConstraintApproval(deliveryConfig.name, environment.name, v, artifact.reference)
+                      if (constraintRunner.checkStatelessConstraints(artifact, deliveryConfig, artifactVersion.version, environment)) {
+                        approveVersion(deliveryConfig, artifact, artifactVersion.version, environment.name)
+                        repository.deleteArtifactVersionQueuedForApproval(
+                          deliveryConfig.name, environment.name, artifactVersion.version, artifact.reference)
+                      } else {
+                        log.debug("Version ${artifactVersion.version} of $artifact does not currently pass stateless constraints")
+                        queuedForApproval.remove(artifactVersion)
```

Comment:
This was missing in the code before, and would affect the logging statements down below. No effect on behavior, just that we might get misleading logs otherwise.